core/src/main/java/org/apache/iceberg/data/ContentFileWriteBuilderImpl.java

```diff
@@ -0,0 +1,336 @@
+ * specialized writer for the requested content type.
+ *
+ * @param <B> the concrete builder type for method chaining
+ * @param <S> the type of the schema for the input data
+ * @param <D> the type of data records the writer will accept
+ */
+@SuppressWarnings("unchecked")
+abstract class ContentFileWriteBuilderImpl<B extends ContentFileWriteBuilder<B, S>, D, S>
+    implements ContentFileWriteBuilder<B, S> {
+  private final WriteBuilder<D, S> writeBuilder;
+  private final String location;
+  private final FileFormat format;
+  private PartitionSpec spec = null;
+  private StructLike partition = null;
+  private EncryptionKeyMetadata keyMetadata = null;
+  private SortOrder sortOrder = null;
+
+  static <D, S> DataWriteBuilder<D, S> forDataFile(
+      WriteBuilder<D, S> writeBuilder, String location, FileFormat format) {
+    return new DataFileWriteBuilder<>(writeBuilder, location, format);
+  }
+
+  static <D, S> EqualityDeleteWriteBuilder<D, S> forEqualityDelete(
+      WriteBuilder<D, S> writeBuilder, String location, FileFormat format) {
+    return new EqualityDeleteFileWriteBuilder<>(writeBuilder, location, format);
+  }
+
+  static <D, S> PositionDeleteWriteBuilder<D, S> forPositionDelete(
+      WriteBuilder<PositionDelete<D>, S> writeBuilder, String location, FileFormat format) {
+    return new PositionDeleteFileWriteBuilder<>(writeBuilder, location, format);
+  }
+
+  private ContentFileWriteBuilderImpl(
+      WriteBuilder<D, S> writeBuilder, String location, FileFormat format) {
+    this.writeBuilder = writeBuilder;
+    this.location = location;
+    this.format = format;
+  }
+
+  @Override
+  public B set(String property, String value) {
+    writeBuilder.set(property, value);
+    return self();
+  }
+
+  @Override
+  public B meta(String property, String value) {
+    writeBuilder.meta(property, value);
+    return self();
+  }
+
+  @Override
+  public B metricsConfig(MetricsConfig metricsConfig) {
+    writeBuilder.metricsConfig(metricsConfig);
+    return self();
+  }
+
+  @Override
+  public B overwrite() {
+    writeBuilder.overwrite();
+    return self();
+  }
+
+  @Override
+  public B fileEncryptionKey(ByteBuffer encryptionKey) {
+    writeBuilder.fileEncryptionKey(encryptionKey);
+    return self();
+  }
+
+  @Override
+  public B fileAADPrefix(ByteBuffer aadPrefix) {
+    writeBuilder.fileAADPrefix(aadPrefix);
+    return self();
+  }
+
+  @Override
+  public B spec(PartitionSpec newSpec) {
+    this.spec = newSpec;
+    return self();
+  }
+
+  @Override
+  public B partition(StructLike newPartition) {
+    this.partition = newPartition;
+    return self();
+  }
+
+  @Override
+  public B keyMetadata(EncryptionKeyMetadata newKeyMetadata) {
+    this.keyMetadata = newKeyMetadata;
+    return self();
+  }
+
+  @Override
+  public B sortOrder(SortOrder newSortOrder) {
+    this.sortOrder = newSortOrder;
+    return self();
+  }
+
+  private static class DataFileWriteBuilder<D, S>
```

Comment:
Both `ContentFileWriteBuilderImpl` and `DataWriteBuilder` extends from or implements `ContentFileWriteBuilderImpl` interface. wondering if it is a sign that if the hierarchy can be improved.

E.g., one solution is to have `DataWriteBuilder` extends from the abstract class `ContentFileWriteBuilderImpl`. the down side is that  `DataWriteBuilder` couldn't be an interface anymore, which is also non ideal.

Reply:
`DataWriteBuilder`, `EqualityDeleteWriteBuilder` and `PositionDeleteWriteBuilder` are intended as public interfaces which we expose. The `ContentFileWriteBuilder` is only there for the common parts, so we don't have duplicate codes.

`ContentFileWriteBuilderImpl` implements the `ContentFileWriteBuilder` part of the APIs, but not intended as public. The other classes implement the specfic parts, but only exposed through the interfaces, and not as classes themselves.

Does this makes sense?

I would want to keep the `DataWriteBuilder`, `EqualityDeleteWriteBuilder` and `PositionDeleteWriteBuilder`. We might remove the `ContentFileWriteBuilder` interface but that only adds back duplicated code, so I wouldn't do that.

Thanks,
Peter